lambda/transl_array_comprehension.ml

```diff
@@ -0,0 +1,932 @@
+    | Dynamic_size, Pfloatarray ->
+        Mutable, Resizable_array.make ~loc array_kind (float 0.)
+  in
+  Let_binding.make_id array_let_kind Pgenval "array" array_value
+
+(** Generate the code for the body of an array comprehension.  This involves
+    translating the body expression (a [Typedtree.expression], which is the
+    [comp_body] field of [Typedtree.comprehension), but also handles the logic
+    of filling in the array that is being produced by the comprehension.  This
+    logic varies depending on whether we are subject to the float array
+    optimization or not and whether we are in the fixed size array case or not,
+    so the correctness depends on getting the correct bindings from
+    [initial_array] and [clauses]. *)
+let body
+      ~loc
+      ~array_kind
+      ~array_size:{array_size; array_sizing}
+      ~array
+      ~index
+      ~body
+  =
+  (* The body of an array comprehension has three jobs:
+       1. Compute the next element
+       2. Assign it (mutably) to the next element of the array
+       3. Advance the index of the next element
+     However, there are several pieces of complexity:
+       (a) If the array size is not fixed, we have to check if the index has
+           overflowed; if it has, we have to double the size of the array.  (The
+           complex case corresponds to [array_size.array_sizing] being
+           [Dynamic_size].)
+       (b) If the array kind is not statically known, we initially created an
+           empty array; we have to check if we're on the first iteration and use
+           the putative first element of the array as the placeholder value for
+           every element of the array.  (The complex case corresponds to
+           [array_kind] being [Pgenarray].)
+       (c) If both (a) and (b) hold, we shouldn't bother checking for an
+           overflowed index on the first loop iteration.
+     The result is that we build the "set the element" behavior in three steps:
+       i.   First, we build the raw "set the element unconditionally" expression
+            ([set_element_raw]).
+       ii.  Then, if necessary, we precede that with the resizing check;
+            otherwise, we leave the raw behavior alone
+            ([set_element_in_bounds]).
+       iii. Then, if necessary, we check to see if we're on the first iteration
+            and create the fresh array instead if so; otherwise, we leave the
+            size-safe behavior alone ([set_element_known_kind_in_bounds]).
+       iv.  Finally, we take the resulting safe element-setting behavior (which
+            could be equal to the result from any of stages i--iii), and follow
+            it up by advancing the index of the element to update.
+  *)
+  let open (val Lambda_utils.int_ops ~loc) in
+  let set_element_raw elt =
+    (* array.(index) <- elt *)
+    Lprim(Parraysetu array_kind, [Lvar array; Lvar index; elt], loc)
+      (* CR aspectorzabusky: Is [array_kind] safe here, since it could be
```

Comment:
In `Lambda`, even if dealing with an array of unboxed floats (kind `Pfloatarray`), the arguments and results of the get/set primitives are still boxed.  I'm not exactly sure what the comment is asking here, but if `elt` is just a normal OCaml value (which might be a boxed float, to be put in an array of unboxed floats), it should be fine.

The way I reminded myself of this semantics, by the way, is to look [here](https://github.com/ocaml-flambda/flambda-backend/blob/main/middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml#L501).  This file decomposes the majority of `Lambda` primitives into unboxed operations, so it's a good reference.

Reply:
I think I was asking if it was safe to have `Parraysetu Pgenarray` in general, since `Pgenarray` requires checking what kind of things are in the array.  It sounds like `Parraysetu` handles that checking itself, though?